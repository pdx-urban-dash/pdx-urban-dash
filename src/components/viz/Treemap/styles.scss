/*
These are the global variables we use. Just an optimization to keep
us using standardized widths and colors. Reference base-styles.scss when
defining font-families, sizes, margins, padding or colors.
*/
@import "../../../base-styles";

.LineChart {
  &-wrapper {
    height: 100%;
    width: 100%;
  }

  &-title {
    font-family: $font-heading;
    font-size: $font-size-lg;
    margin-left: auto;
    margin-right: auto;
  }

  &-theme {
    &__light {
      color: $color-font-dark;
      background-color: white;
    }

    &__dark {
      color: $color-font-light;
      background-color: $color-font-dark;
    }
  }
}
