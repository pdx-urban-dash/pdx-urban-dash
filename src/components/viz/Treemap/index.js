import React from 'react';
import PropTypes from 'prop-types';
import { createClassFromSpec } from 'react-vega';

import { injectPropsIntoSchema } from '../../../utils/vegaUtils';
import { getDefaultColor } from '../../../utils/vizUtils';
import { colors, sizes } from '../../../constants';
import TreemapSchema from './schema';
import './styles.scss';

import Legend from '../shared/Legend';
import ChartWrapper from '../shared/ChartWrapper';

const buildTreemapData = (data) => {
  const retData = [{
    name: 'City Of Portland',
    id: 'City Of Portland',
  }];
  data.forEach((dataSet) => {
    const bureauName = dataSet.Bureau;
    const serviceName = dataSet['City Service Area'];
    const councilName = dataSet['Council Priority Area'];
    const measureName = dataSet['Measure Title'];
    const measureValue = dataSet['FY 2017-18 Actual'];

    const bureauID = bureauName;
    const serviceID = `${bureauName}, ${serviceName}`;
    const councilID = `${serviceID}, ${serviceName}`;
    const measureID = `${councilID}, ${measureName}`;

    if (retData.findIndex(x => x.id === councilID) === -1) {
      retData.push({
        name: councilName,
        id: councilID,
        parent: serviceID,
      });

      if (retData.findIndex(x => x.id === serviceID) === -1) {
        retData.push({
          name: serviceName,
          id: serviceID,
          parent: bureauID,
        });

        if (retData.findIndex(x => x.id === bureauID) === -1) {
          retData.push({
            id: bureauID,
            name: bureauName,
            parent: 'City Of Portland',
          });
        }
      }
    }

    retData.push({
      name: measureName,
      id: measureID,
      parent: councilID,
      size: measureValue,
    });
  });
  console.log(retData);
  return retData;
};

const getColors = data => data.map((elem, idx) => elem.barColor || getDefaultColor(idx));

const Treemap = (
  {
    data,
    title,
    className,
    colorScheme,
  },
) => {
  const colorRange = getColors(data);
  const Graph = createClassFromSpec(
    injectPropsIntoSchema(
      {
        $dataValues: {
          name: 'values',
          value: buildTreemapData(data),
        },
        $labelColor: {
          name: 'labelColor',
          value: colorScheme === 'light' ? colors.fontDark : colors.fontLight,
        },
        $labelFontSize: {
          name: 'labelFontSize',
          value: parseInt(sizes.fontSM.slice(0, -2), 10),
        },
        $colorRange: {
          name: 'range',
          value: colorRange,
        },
      },
      TreemapSchema,
    ),
  );
  return (
    <ChartWrapper title={title} className={className}>
      <Graph />
      <Legend
        series={data.map((elem, idx) => ({
          title: elem.dataSetName,
          description: elem.description,
          color: elem.barColor || getDefaultColor(idx),
        }))}
      />
    </ChartWrapper>
  );
};

Treemap.propTypes = {
  data: PropTypes.arrayOf(PropTypes.shape({
    dataSetName: PropTypes.string,
    barColor: PropTypes.string,
    description: PropTypes.string,
    values: PropTypes.arrayOf(PropTypes.shape({
      x: PropTypes.number.isRequired,
      y: PropTypes.number.isRequired,
    })),
  })).isRequired,
  title: PropTypes.string,
  className: PropTypes.string,
  colorScheme: PropTypes.oneOf(['light', 'dark']),
};

Treemap.defaultProps = {
  title: 'chart title',
  className: '',
  colorScheme: 'light',
};

export default Treemap;
