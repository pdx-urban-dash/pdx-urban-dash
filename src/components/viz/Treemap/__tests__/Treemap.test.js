import React from 'react';
import { shallow } from 'enzyme';
import Treemap from '../index';

const testData = [
  {
    'Measure Title': 'Number of Building Permits',
    'Council Priority Area': 'Housing & Homelessness',
    'City Service Area': 'Community Development',
    Bureau: 'Development Services',
    'FY 2016-17 Actual': '  8,145 ',
    'FY 2017-18 Actual': '  6,905 ',
    'Desired Trend': 'UP',
    'Strategic Target': '  8,100 ',
    'What the measure shows': 'Construction market activity',
    Notes: 'Data can be disagregated into commercial and residential, or even further into permit type (new construction, renovation, etc.)',
    'What is important about this metric?': '',
    'Data frequency': '',
    'Data Source': '',
    'Data Visualization': '',
    'Geographic Tag?': 'y',
    Target: '',
  },
  {
    'Measure Title': 'Percentage of 9-1-1 calls answered within 20 seconds',
    'Council Priority Area': 'Public Safety',
    'City Service Area': 'Public Safety',
    Bureau: 'Bureau of Emergency Communications',
    'FY 2016-17 Actual': '63.2%',
    'FY 2017-18 Actual': '63.9%',
    'Desired Trend': 'UP',
    'Strategic Target': '90.0%',
    'What the measure shows': 'Speed by which we answer 9-1-1 calls',
    Notes: 'Can be updated daily, can be disaggregated into land line vs. cell phone calls. ',
    'What is important about this metric?': '',
    'Data frequency': '',
    'Data Source': 'BOEC',
    'Data Visualization': '',
    'Geographic Tag?': 'n',
    Target: '',
  },
];

const testTitle = 'chart title';

describe('<Treemap />', () => {
  test('snapshot', () => {
    const wrapper = shallow(<Treemap data={testData} title={testTitle} />);
    expect(wrapper).toMatchSnapshot();
  });
});
