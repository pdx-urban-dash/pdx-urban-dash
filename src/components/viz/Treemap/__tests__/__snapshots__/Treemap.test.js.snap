// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Treemap /> snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Treemap
    className=""
    colorScheme="light"
    data={
      Array [
        Object {
          "Bureau": "Development Services",
          "City Service Area": "Community Development",
          "Council Priority Area": "Housing & Homelessness",
          "Data Source": "",
          "Data Visualization": "",
          "Data frequency": "",
          "Desired Trend": "UP",
          "FY 2016-17 Actual": "  8,145 ",
          "FY 2017-18 Actual": "  6,905 ",
          "Geographic Tag?": "y",
          "Measure Title": "Number of Building Permits",
          "Notes": "Data can be disagregated into commercial and residential, or even further into permit type (new construction, renovation, etc.)",
          "Strategic Target": "  8,100 ",
          "Target": "",
          "What is important about this metric?": "",
          "What the measure shows": "Construction market activity",
        },
        Object {
          "Bureau": "Bureau of Emergency Communications",
          "City Service Area": "Public Safety",
          "Council Priority Area": "Public Safety",
          "Data Source": "BOEC",
          "Data Visualization": "",
          "Data frequency": "",
          "Desired Trend": "UP",
          "FY 2016-17 Actual": "63.2%",
          "FY 2017-18 Actual": "63.9%",
          "Geographic Tag?": "n",
          "Measure Title": "Percentage of 9-1-1 calls answered within 20 seconds",
          "Notes": "Can be updated daily, can be disaggregated into land line vs. cell phone calls. ",
          "Strategic Target": "90.0%",
          "Target": "",
          "What is important about this metric?": "",
          "What the measure shows": "Speed by which we answer 9-1-1 calls",
        },
      ]
    }
    title="chart title"
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": Array [
        <a />,
        <Legend
          series={
            Array [
              Object {
                "color": "#82204A",
                "description": undefined,
                "title": undefined,
              },
              Object {
                "color": "#558C8C",
                "description": undefined,
                "title": undefined,
              },
            ]
          }
        />,
      ],
      "className": "",
      "title": "chart title",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {},
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "series": Array [
            Object {
              "color": "#82204A",
              "description": undefined,
              "title": undefined,
            },
            Object {
              "color": "#558C8C",
              "description": undefined,
              "title": undefined,
            },
          ],
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": Array [
          <a />,
          <Legend
            series={
              Array [
                Object {
                  "color": "#82204A",
                  "description": undefined,
                  "title": undefined,
                },
                Object {
                  "color": "#558C8C",
                  "description": undefined,
                  "title": undefined,
                },
              ]
            }
          />,
        ],
        "className": "",
        "title": "chart title",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {},
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "series": Array [
              Object {
                "color": "#82204A",
                "description": undefined,
                "title": undefined,
              },
              Object {
                "color": "#558C8C",
                "description": undefined,
                "title": undefined,
              },
            ],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromError": true,
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
